### ISSUE TITLE
vmstorage statefulset  not correct update  when the container of vmBackup  has extra "vmstorage-db" mount

### ISSUE THREAD

**Issue by wu0407** [VictoriaMetrics/operator#366]
If vmBackup container has extra `vmstorage-db` mount path,  next remove this extra mount in vmcluster,  the result is that remain old  `vmstorage-db` mount in statefulset  when upgrade from 0.19.0 to 0.20.1 

old vmcluster:
```
 vmstorage:
    extraArgs:
      search.maxUniqueTimeseries: "600000"
    image:
      tag: v1.67.0-cluster
    replicaCount: 3
    resources:
      limits:
        cpu: "4"
        memory: 16Gi
      requests:
        cpu: 50m
        memory: 1Gi
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 100Gi
          storageClassName: cbs-ssd-prepaid
        status: {}
    storageDataPath: /vm-data
    vmBackup:
      acceptEULA: true
      credentialsSecret:
        key: credentials-backup.yaml
        name: credentials-backup
      customS3Endpoint: https://cos.ap-shanghai.myqcloud.com
      destination: s3://test-thanos/$(NODENAME)
      disableMonthly: true
      disableWeekly: true
      extraArgs:
        keepLastDaily: "30"
        keepLastHourly: "72"
        runOnStart: "true"
      extraEnvs:
      - name: AWS_DEFAULT_REGION
        value: ap-shanghai
      - name: NODENAME
        valueFrom:
          fieldRef:
            fieldPath: metadata.name
      volumeMounts:
      - mountPath: /vm-data
        name: vmstorage-db
```
old vmstorage statefulset:
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  annotations:
    meta.helm.sh/release-name: victoria-metrics-k8s-stack
    meta.helm.sh/release-namespace: victoria
  creationTimestamp: "2021-11-01T06:24:32Z"
  finalizers:
  - apps.victoriametrics.com/finalizer
  generation: 3
  labels:
    app.kubernetes.io/component: monitoring
    app.kubernetes.io/instance: victoria-metrics-k8s-stack
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: vmstorage
    app.kubernetes.io/version: 1.67.0
    helm.sh/chart: victoria-metrics-k8s-stack-0.5.4
    managed-by: vm-operator
  name: vmstorage-victoria-metrics-k8s-stack
  namespace: victoria
  ownerReferences:
  - apiVersion: operator.victoriametrics.com/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: VMCluster
    name: victoria-metrics-k8s-stack
    uid: 4073bd0e-4a12-4afb-b85f-90776685cca0
  resourceVersion: "333641442"
  selfLink: /apis/apps/v1/namespaces/victoria/statefulsets/vmstorage-victoria-metrics-k8s-stack
  uid: c25e0965-f8b5-4a5a-aab8-dd3018ca39aa
spec:
  podManagementPolicy: Parallel
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/component: monitoring
      app.kubernetes.io/instance: victoria-metrics-k8s-stack
      app.kubernetes.io/name: vmstorage
      managed-by: vm-operator
  serviceName: vmstorage-victoria-metrics-k8s-stack
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: monitoring
        app.kubernetes.io/instance: victoria-metrics-k8s-stack
        app.kubernetes.io/name: vmstorage
        managed-by: vm-operator
    spec:
      containers:
      - args:
        - -dedup.minScrapeInterval=1ms
        - -httpListenAddr=:8482
        - -retentionPeriod=14d
        - -search.maxUniqueTimeseries=600000
        - -storageDataPath=/vm-data
        - -vminsertAddr=:8400
        - -vmselectAddr=:8401
        image: victoriametrics/vmstorage:v1.67.0-cluster
        imagePullPolicy: IfNotPresent
        name: vmstorage
        ports:
        - containerPort: 8482
          name: http
          protocol: TCP
        - containerPort: 8400
          name: vminsert
          protocol: TCP
        - containerPort: 8401
          name: vmselect
          protocol: TCP
        readinessProbe:
          failureThreshold: 10
          httpGet:
            path: /health
            port: 8482
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          limits:
            cpu: "4"
            memory: 16Gi
          requests:
            cpu: 50m
            memory: 1Gi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: FallbackToLogsOnError
        volumeMounts:
        - mountPath: /vm-data
          name: vmstorage-db
      - args:
        - -credsFilePath=/etc/vm/creds/credentials-backup.yaml
        - -customS3Endpoint=https://cos.ap-shanghai.myqcloud.com
        - -disableMonthly
        - -disableWeekly
        - -dst=s3://test-thanos/$(NODENAME)
        - -envflag.enable=true
        - -eula
        - -keepLastDaily=30
        - -keepLastHourly=72
        - -runOnStart=true
        - -snapshot.createURL=http://localhost:8482/snapshot/create
        - -snapshot.deleteURL=http://localhost:8482/snapshot/delete
        - -storageDataPath=/vm-data
        env:
        - name: AWS_DEFAULT_REGION
          value: ap-shanghai
        - name: NODENAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        image: victoriametrics/vmbackupmanager:v1.66.2-enterprise
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /health
            port: 8300
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 5
        name: vmbackuper
        ports:
        - containerPort: 8300
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 10
          httpGet:
            path: /health
            port: 8300
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          limits:
            cpu: 500m
            memory: 500Mi
          requests:
            cpu: 150m
            memory: 200Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: FallbackToLogsOnError
        volumeMounts:
        - mountPath: /victoria-metrics-data
          name: vmstorage-db
          readOnly: true
        - mountPath: /vm-data
          name: vmstorage-db
        - mountPath: /etc/vm/creds
          name: secret-credentials-backup
          readOnly: true
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: vmcluster-victoria-metrics-k8s-stack
      serviceAccountName: vmcluster-victoria-metrics-k8s-stack
      terminationGracePeriodSeconds: 30
      volumes:
      - name: secret-credentials-backup
        secret:
          defaultMode: 420
          secretName: credentials-backup
  updateStrategy:
    type: OnDelete
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      creationTimestamp: null
      name: vmstorage-db
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 100Gi
      storageClassName: cbs-ssd-prepaid
      volumeMode: Filesystem
    status:
      phase: Pending
```

new vmcluster:
```
spec:
  replicationFactor: 2
  retentionPeriod: 14d
  vminsert:
    image:
      tag: v1.68.0-cluster
    replicaCount: 2
    resources:
      limits:
        cpu: "2"
        memory: 1000Mi
      requests:
        cpu: 500m
        memory: 500Mi
  vmselect:
    cacheMountPath: /select-cache
    extraArgs:
      search.maxQueryDuration: 120s
    image:
      tag: v1.68.0-cluster
    replicaCount: 2
    resources:
      limits:
        cpu: "4"
        memory: 2Gi
      requests:
        cpu: 500m
        memory: 512Mi
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 20Gi
          storageClassName: cbs-ssd-prepaid
        status: {}
  vmstorage:
    extraArgs:
      search.maxUniqueTimeseries: "600000"
    image:
      tag: v1.68.0-cluster
    replicaCount: 3
    resources:
      limits:
        cpu: "4"
        memory: 16Gi
      requests:
        cpu: 50m
        memory: 1Gi
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 100Gi
          storageClassName: cbs-ssd-prepaid
        status: {}
    storageDataPath: /vm-data
    vmBackup:
      acceptEULA: true
      credentialsSecret:
        key: credentials-backup.yaml
        name: credentials-backup
      customS3Endpoint: https://cos.ap-shanghai.myqcloud.com
      destination: s3://test-thanos
      disableMonthly: true
      disableWeekly: true
      extraArgs:
        keepLastDaily: "30"
        keepLastHourly: "72"
        runOnStart: "true"
      extraEnvs:
      - name: AWS_DEFAULT_REGION
        value: ap-shanghai
```
new vmstorage statfulset:
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  annotations:
    meta.helm.sh/release-name: victoria-metrics-k8s-stack
    meta.helm.sh/release-namespace: victoria
  creationTimestamp: "2021-11-01T06:24:32Z"
  finalizers:
  - apps.victoriametrics.com/finalizer
  generation: 4
  labels:
    app.kubernetes.io/component: monitoring
    app.kubernetes.io/instance: victoria-metrics-k8s-stack
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: vmstorage
    app.kubernetes.io/version: 1.68.0
    helm.sh/chart: victoria-metrics-k8s-stack-0.5.7
    managed-by: vm-operator
  name: vmstorage-victoria-metrics-k8s-stack
  namespace: victoria
  ownerReferences:
  - apiVersion: operator.victoriametrics.com/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: VMCluster
    name: victoria-metrics-k8s-stack
    uid: 4073bd0e-4a12-4afb-b85f-90776685cca0
  resourceVersion: "333647449"
  selfLink: /apis/apps/v1/namespaces/victoria/statefulsets/vmstorage-victoria-metrics-k8s-stack
  uid: c25e0965-f8b5-4a5a-aab8-dd3018ca39aa
spec:
  podManagementPolicy: Parallel
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/component: monitoring
      app.kubernetes.io/instance: victoria-metrics-k8s-stack
      app.kubernetes.io/name: vmstorage
      managed-by: vm-operator
  serviceName: vmstorage-victoria-metrics-k8s-stack
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: monitoring
        app.kubernetes.io/instance: victoria-metrics-k8s-stack
        app.kubernetes.io/name: vmstorage
        managed-by: vm-operator
    spec:
      containers:
      - args:
        - -dedup.minScrapeInterval=1ms
        - -httpListenAddr=:8482
        - -retentionPeriod=14d
        - -search.maxUniqueTimeseries=600000
        - -storageDataPath=/vm-data
        - -vminsertAddr=:8400
        - -vmselectAddr=:8401
        image: victoriametrics/vmstorage:v1.68.0-cluster
        imagePullPolicy: IfNotPresent
        name: vmstorage
        ports:
        - containerPort: 8482
          name: http
          protocol: TCP
        - containerPort: 8400
          name: vminsert
          protocol: TCP
        - containerPort: 8401
          name: vmselect
          protocol: TCP
        readinessProbe:
          failureThreshold: 10
          httpGet:
            path: /health
            port: 8482
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          limits:
            cpu: "4"
            memory: 16Gi
          requests:
            cpu: 50m
            memory: 1Gi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: FallbackToLogsOnError
        volumeMounts:
        - mountPath: /vm-data
          name: vmstorage-db
      - args:
        - -credsFilePath=/etc/vm/creds/credentials-backup.yaml
        - -customS3Endpoint=https://cos.ap-shanghai.myqcloud.com
        - -disableMonthly
        - -disableWeekly
        - -dst=s3://test-thanos
        - -envflag.enable=true
        - -eula
        - -keepLastDaily=30
        - -keepLastHourly=72
        - -runOnStart=true
        - -snapshot.createURL=http://localhost:8482/snapshot/create
        - -snapshot.deleteURL=http://localhost:8482/snapshot/delete
        - -storageDataPath=/vm-data
        env:
        - name: AWS_DEFAULT_REGION
          value: ap-shanghai
        image: victoriametrics/vmbackupmanager:v1.66.2-enterprise
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /health
            port: 8300
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 5
        name: vmbackuper
        ports:
        - containerPort: 8300
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 10
          httpGet:
            path: /health
            port: 8300
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          limits:
            cpu: 500m
            memory: 500Mi
          requests:
            cpu: 150m
            memory: 200Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: FallbackToLogsOnError
        volumeMounts:
        - mountPath: /victoria-metrics-data
          name: vmstorage-db
          readOnly: true
        - mountPath: /etc/vm/creds
          name: secret-credentials-backup
          readOnly: true
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: vmcluster-victoria-metrics-k8s-stack
      serviceAccountName: vmcluster-victoria-metrics-k8s-stack
      terminationGracePeriodSeconds: 30
      volumes:
      - name: secret-credentials-backup
        secret:
          defaultMode: 420
          secretName: credentials-backup
  updateStrategy:
    type: OnDelete
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      creationTimestamp: null
      name: vmstorage-db
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 100Gi
      storageClassName: cbs-ssd-prepaid
      volumeMode: Filesystem
    status:
      phase: Pending
```

**Comment by wu0407** [comment]
relate logs in operator:
```
{"level":"info","ts":1635752221.7648149,"logger":"factory","msg":"create or update vmstorage for cluster","controller":"vmstorage","cluster":"victoria-metrics-k8s-stack"}
{"level":"info","ts":1635752221.7701364,"logger":"factory","msg":"vmstorage was found, updating it","controller":"vmstorage","cluster":"victoria-metrics-k8s-stack"}
{"level":"info","ts":1635752223.7734144,"logger":"client_utils","msg":"sts update is needed","sts":"vmstorage-victoria-metrics-k8s-stack","currentVersion":"vmstorage-victoria-metrics-k8s-stack-64c699b6db","desiredVersion":"vmstorage-victoria-metrics-k8s-stack-d8bdd7f48"}
{"level":"info","ts":1635752223.7734485,"logger":"client_utils","msg":"checking if update needed","controller":"sts.rollingupdate","desiredVersion":"vmstorage-victoria-metrics-k8s-stack-d8bdd7f48","wasRecreated":false}
{"level":"info","ts":1635752223.777197,"logger":"client_utils","msg":"starting pods update, checking updated, by not ready pods","controller":"sts.rollingupdate","desiredVersion":"vmstorage-victoria-metrics-k8s-stack-d8bdd7f48","wasRecreated":false,"updated pods count":1,"desired version":"vmstorage-victoria-metrics-k8s-stack-d8bdd7f48"}
{"level":"info","ts":1635752223.777214,"logger":"client_utils","msg":"checking ready status for already updated pod to desired version","controller":"sts.rollingupdate","desiredVersion":"vmstorage-victoria-metrics-k8s-stack-d8bdd7f48","wasRecreated":false,"pod":"vmstorage-victoria-metrics-k8s-stack-0"}
{"level":"error","ts":1635752313.7865117,"logger":"client_utils","msg":"cannot get ready status for already updated pod","controller":"sts.rollingupdate","desiredVersion":"vmstorage-victoria-metrics-k8s-stack-d8bdd7f48","wasRecreated":false,"pod":"vmstorage-victoria-metrics-k8s-stack-0","error":"timed out waiting for the condition","stacktrace":"github.com/VictoriaMetrics/operator/controllers/factory/k8stools.HandleSTSUpdate\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/k8stools/sts.go:37\ngithub.com/VictoriaMetrics/operator/controllers/factory.createOrUpdateVMStorage\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/vmcluster.go:396\ngithub.com/VictoriaMetrics/operator/controllers/factory.CreateOrUpdateVMCluster\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/vmcluster.go:95\ngithub.com/VictoriaMetrics/operator/controllers.(*VMClusterReconciler).Reconcile\n\tgithub.com/VictoriaMetrics/operator/controllers/vmcluster_controller.go:68\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.9.0/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.9.0/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.9.0/pkg/internal/controller/controller.go:214"}
```

**Comment by wu0407** [comment]
I found out the reason is use helm chart simultaneous update operator and vmbackup. 0.19.0 operator first handle vmbackup update, then 0.21.0 operator second handle update, but it stuck in pod crashLoop and then context timeout.

https://github.com/VictoriaMetrics/operator/blob/a34b6620c70ef9f2b26ab2ff92d5cba4281b5f5e/controllers/factory/k8stools/sts.go#L137

https://github.com/VictoriaMetrics/operator/blob/a34b6620c70ef9f2b26ab2ff92d5cba4281b5f5e/controllers/factory/k8stools/sts.go#L202-L226

**Comment by f41gh7** [comment]
Sorry for delay, it must be fixed at related PR. It was a regression at statefulset update mechanism. Can you verify it with docker image: `victoriametrics/operator:gh-366` ?
 
 Thanks for investigation, it helps a lot!

**Comment by f41gh7** [comment]
Changes was included to the 0.21.0 release.
